functions-python/feed_sync_process_transitland/src/main.py

```diff
@@ -0,0 +1,509 @@
+        """
+        try:
+            logger.info(
+                f"Starting feed processing "
+                f"for external_id: {payload.external_id}"
+            )
+
+            # Check current state in database
+            current_feed_id, current_url = self.get_current_feed_info(
+                payload.external_id, payload.source
+            )
+
+            if current_feed_id is None:
+                logger.info("Processing new feed")
+                # If no existing feed found - checks if URL exists in any feed
+                if self.check_feed_url_exists(payload.feed_url):
+                    logger.info(f"Feed URL already exists: {payload.feed_url}")
+                    return
+                # Create new feed
+                self.process_new_feed(payload)
+            else:
+                # If Feed exists - check if URL has changed
+                if current_url != payload.feed_url:
+                    logger.info("Processing feed update")
+                    logger.debug(
+                        f"Found existing feed: "
+                        f"{current_feed_id} with different URL"
+                    )
+                    # URL changed - handle update
+                    self.process_feed_update(payload, current_feed_id)
+                else:
+                    logger.info(
+                        f"Feed already exists with "
+                        f"same URL: {payload.external_id}"
+                    )
+                    return
+
+            self.session.commit()
+            logger.debug("Database transaction committed successfully")
+
+            # Publish to dataset batch topic if not authenticated
+            if not payload.auth_info_url:
+                self.publish_to_batch_topic(payload)
+
+        except Exception as e:
+            error_msg = (f"Error processing "
+                         f"feed {payload.external_id}: {str(e)}")
+            logger.error(error_msg)
+            self.session.rollback()
+            logger.debug("Database transaction rolled back due to error")
+            raise
+
+    def process_new_feed(self, payload: FeedPayload) -> None:
+        """
+        Process creation of a new feed
+
+        Args:
+            payload (FeedPayload): The feed payload for new feed
+        """
+        logger.info(
+            f"Starting new feed creation "
+            f"for external_id: {payload.external_id}"
+        )
+
+        # Checks if feed with same URL exists
+        if self.check_feed_url_exists(payload.feed_url):
+            logger.info(f"Feed URL already exists: {payload.feed_url}")
+            return
+
+        # Generate new feed ID and stable ID
+        feed_id = str(uuid.uuid4())
+        stable_id = f"{payload.source}-{payload.external_id}"
+
+        logger.debug(f"Generated new feed_id: "
+                     f"{feed_id} and stable_id: {stable_id}")
+
+        try:
+            # Insert new feed
+            feed_query = text(
```

Comment:
Let's use SqlAlchemy instead of plain text queries in all cases(insert, update, and select).  SqlAchemy provides better maintainability, readability, and security compared to plain SQL. It also simplifies migrations, refactoring, and testing. If you need inspiration for the implementation, check [populate_db_gtfs.py](https://github.com/MobilityData/mobility-feed-api/blob/main/api/src/scripts/populate_db_gtfs.py)